jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java

```diff
@@ -0,0 +1,457 @@
+
+            baseRequest.setAttribute(HttpConnection.UPGRADE_CONNECTION_ATTRIBUTE, connection);
+            return true;
+        }
+    };
+
+
+
+    static class TestWebSocketConnectionFactory extends ContainerLifeCycle implements WebSocketConnectionFactory
+    {
+        WebSocketPolicy policy = new WebSocketPolicy(WebSocketBehavior.SERVER);
+        ByteBufferPool bufferPool = new MappedByteBufferPool();
+        WebSocketExtensionRegistry extensionRegistry = new WebSocketExtensionRegistry();
+        DecoratedObjectFactory objectFactory = new DecoratedObjectFactory();
+
+        @Override
+        public WebSocketCoreConnection newConnection(Connector connector, EndPoint endPoint, List<String> extensions)
+        {
+            ExtensionStack extensionStack = new ExtensionStack(extensionRegistry);
+            extensionStack.negotiate(objectFactory, policy, bufferPool,
+                                     extensions.stream().map(ExtensionConfig::parse).collect(Collectors.toList()));
+
+            UpgradeRequest upgradeRequest = null; // TODO
+            UpgradeResponse upgradeResponse = null; // TODO
+
+            return new WebSocketCoreConnection(
+                    endPoint,
+                    connector.getExecutor(),
+                    bufferPool,
+                    objectFactory,
+                    policy,
+                    extensionStack,
+                    upgradeRequest,
+                    upgradeResponse);
+        }
+    }
+
+
+    static class TestWebSocketSessionFactory extends ContainerLifeCycle implements WebSocketSessionFactory
+    {
+        @Override
+        public WebSocketCoreSession<ContainerLifeCycle,WebSocketCoreConnection,WebSocketLocalEndpoint,WebSocketRemoteEndpoint>
+        newSession(WebSocketCoreConnection connection, String subprotocol)
+        {
+
+            // TODO why is remoteEndpoint an interface rather than just an impl?
+            WebSocketRemoteEndpointImpl remoteEndpoint = new WebSocketRemoteEndpointImpl(connection);
+            remoteEndpoint.open(); // TODO why is this needed?
+
+            // TODO abstract the creation of a local Endpoint
+            WebSocketLocalEndpoint localEndpoint = new WebSocketLocalEndpoint()
+            {
+                @Override
+                public Logger getLog()
+                {
+                    return LOG;
+                }
+
+                @Override
+                public boolean isOpen()
+                {
+                    return connection.isOpen();
+                }
+
+                @Override
+                public void onOpen()
+                {
+                    LOG.debug("onOpen {}",this);
+                    TextFrame text = new TextFrame();
+                    text.setPayload("Opened!");
+
+                    // TODO do I need to mask this myself? if so then why?
+                    remoteEndpoint.sendFrame(text, new Callback()
```

Comment:
You should be using Session.getRemote().sendText().

The WebSocketRemoteEndpoint is not an end user API.